langchain/src/tools/ical.ts

```diff
@@ -0,0 +1,104 @@
+import { z } from "zod";
+import { ICalCalendar } from "ical-generator";
+
+import { BaseLLM } from "../llms/base.js";
+import { PromptTemplate } from "../prompts/prompt.js";
+import { Tool, ToolParams } from "./base.js";
+import { StructuredOutputParser } from "../output_parsers/structured.js";
+import { BaseFileStore } from "../schema/index.js";
+
+interface ICalParams extends ToolParams {
+  store: BaseFileStore;
+  llm: BaseLLM;
+}
+
+export class ICalTool extends Tool {
+  name = "iCal";
+
+  description =
+    "Useful for scheduling events and storing it locally as iCal file. It can only create new events, it cannot edit existing events. Input should be raw instruction(s) describing the event to be created in detail.";
+
+  eventsSchema = z.object({
+    name: z.string().describe("Name of the event"),
+    description: z
+      .string()
+      .describe("A detailed description of the created event(s)."),
+    events: z.array(
+      z.object({
+        start: z.string().describe("Start date of the event"),
+        end: z.string().describe("End date of the event"),
+        summary: z.string().describe("Summary of the event"),
+        description: z.string().describe("Description of the event"),
+        location: z
+          .string()
+          .describe("Location of the event if any")
+          .optional(),
+        url: z.string().describe("URL of the event if any").optional(),
+      })
+    ),
+  });
+
+  store: BaseFileStore;
+
+  llm: BaseLLM;
+
+  constructor({ store, llm, verbose, callbacks }: ICalParams) {
+    super(verbose, callbacks);
+    this.store = store;
+    this.llm = llm;
+  }
+
+  /** @ignore */
+  async _call(input: string): Promise<string> {
+    try {
+      const { name, description, events } = await this.getEventsObject(input);
+      const calendar = new ICalCalendar({ name });
+
+      for (const event of events) {
+        calendar.createEvent({
+          start: new Date(event.start),
+          end: new Date(event.end),
+          summary: event.summary,
+          description: event.description,
+          location: event.location,
+          url: event.url,
+        });
+      }
+
+      await this.store.writeFile(`${name}.ics`, calendar.toString());
+
+      return `${description} event created successfully and saved as ${name}.ics.`;
+    } catch (err) {
+      return "Failed to create calendar event.";
```

Comment:
What happens if this errors due to something non-retryable? I suspect the agent may just continually retry.

Might be better to just throw it so that execution stops.